File: RacingChars/cpp/rc.cpp
// racing chars cpp soln by steven a dunn for code eval

#include <fstream>
#include <iostream>
#include <string>

using std::cout;
using std::endl;
using std::ifstream;
using std::string;

char update_direction(int, int);

int main(int argc, char* argv[])
{
    ifstream input_file(argv[1]);
    string line;

    if (input_file)
    {
        char direction = '|';
        int prev = -1;
        while (getline(input_file, line))
        {
            int gate_index = line.find('_');
            if (prev == -1)
                prev = gate_index;

            int checkpoint_index = line.find('C');
            if (checkpoint_index == string::npos)
            {
                direction = update_direction(prev, gate_index);
                line[gate_index] = direction;
                prev = gate_index;
            }
            else
            {
                direction = update_direction(prev, checkpoint_index);
                line[checkpoint_index] = direction;
                prev = checkpoint_index;
            }
            cout << line << endl;
        }
        input_file.close();
    }
    return 0;
}

char update_direction(int prev, int index)
{
    char direction;
    if (index < prev)
        direction = '/';
    else if (index == prev)
        direction = '|';
    else
        direction = '\\';
    return direction;
}

File: RacingChars/Ruby/rc.rb
# racing chars solution in ruby for code eval by steven a dunn

input_file = ARGV[0]

def update_dir(prev, index)
  dir = nil
  if index < prev
    dir = '/'
  elsif index == prev
    dir = '|'
  else
    dir = "\\"
  end
  dir
end

File.open(input_file, "r") do |file|
  dir = '|'
  prev = nil
  while line = file.gets
    gate_index = line.index('_')
    if prev.nil? 
      prev = gate_index 
    end

    checkpoint_index = line.index('C')
    if checkpoint_index.nil? 
      dir = update_dir(prev, gate_index)
      line[gate_index] = dir
      prev = gate_index
    else
      dir = update_dir(prev, checkpoint_index)
      line[checkpoint_index] = dir
      prev = checkpoint_index
    end

    puts line 
  end
end
